### PR TITLE
Make AbortUpgrade panic on failure

### PR DESCRIPTION
**Pull request by chatton**
## Description

closes: #3903

This PR replaces the AbortUpgrade function with a `MustAbortUpgrade` function which panics if an error occurs during the upgrade process.

### Commit Message / Changelog Entry

N/A Feature branch

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

**Review by charleenfei — APPROVED** [review]
nice work, lgtm!

File: modules/core/04-channel/keeper/upgrade.go [hunk 1]
```diff
@@ -802,9 +802,18 @@ func (k Keeper) constructProposedUpgrade(ctx sdk.Context, portID, channelID stri
...
+// This function is expected to always succeed, a panic will occur if an error occurs.
+func (k Keeper) MustAbortUpgrade(ctx sdk.Context, portID, channelID string, err error) {
+	if err := k.abortUpgrade(ctx, portID, channelID, err); err != nil {
+		panic(err)
+	}
+}
+
+// abortUpgrade will restore the channel state and flush status to their pre-upgrade state so that upgrade is aborted.
+// any unnecessary state is deleted. An error receipt is written, and the OnChanUpgradeRestore callback is called.
+func (k Keeper) abortUpgrade(ctx sdk.Context, portID, channelID string, err error) error {
```
Comment (DimitrisJim) [inline modules/core/04-channel/keeper/upgrade.go:816]: do we see `abortUpgrade` ever being used w/o panicking?

File: modules/core/04-channel/keeper/upgrade.go [hunk 2]
```diff
@@ -802,9 +802,18 @@ func (k Keeper) constructProposedUpgrade(ctx sdk.Context, portID, channelID stri
 	}, nil
 }
 
-// AbortUpgrade will restore the channel state and flush status to their pre-upgrade state so that upgrade is aborted.
+// MustAbortUpgrade will restore the channel state and flush status to their pre-upgrade state so that upgrade is aborted.
 // any unnecessary state is deleted. An error receipt is written, and the OnChanUpgradeRestore callback is called.
```
Comment (damiannolan) [inline modules/core/04-channel/keeper/upgrade.go:806]: I think both godocs can be updated to remove the "and the OnChanUpgradeRestore callback is called"

**Review by damiannolan — APPROVED** [review]
LGTM sir! nice one, I think it makes sense to keep the panic in the one location 👍

File: modules/core/04-channel/keeper/upgrade_test.go [hunk 3]
```diff
@@ -1305,10 +1305,10 @@ func (suite *KeeperTestSuite) TestAbortHandshake() {
 
 			tc.malleate()
 
-			err := channelKeeper.AbortUpgrade(suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID, upgradeError)
```
Comment (DimitrisJim) [inline modules/core/04-channel/keeper/upgrade_test.go:1308]: guess we could also define the function as a var and trim down on the `NotPanics/Panics` calls in a similar vein. Leaving to your discretion.

**Review by DimitrisJim — APPROVED** [review]

**Pull request merged**